api/v1alpha1/freight_types_test.go

```diff
@@ -470,147 +468,208 @@ func TestFreightStatus_AddApprovedStage(t *testing.T) {
+				Name   string   `json:"name"`
+				Age    int      `json:"age"`
+				Active bool     `json:"active"`
+				Tags   []string `json:"tags"`
 			}{
 				Name:   "test-user",
 				Age:    30,
 				Active: true,
-				Tags:   []string{"admin", "developer"},
-				Configs: map[string]string{
-					"theme": "dark",
-					"lang":  "en",
-				},
+				Tags:   []string{"admin", "dev"},
 			},
+			expectedJSON: `{"name":"test-user","age":30,"active":true,"tags":["admin","dev"]}`,
 		},
 		{
-			name: "slice of integers",
-			data: []int{1, 2, 3, 4, 5},
+			name: "update existing metadata",
+			initialState: map[string]apiextensionsv1.JSON{
+				"test-key": {Raw: []byte(`"old-value"`)},
+			},
+			key:          "test-key",
+			data:         "new-value",
+			expectedJSON: `"new-value"`,
 		},
 		{
-			name: "nested map",
-			data: map[string]any{
-				"level1": map[string]any{
-					"level2": map[string]string{
-						"key": "value",
-					},
-				},
-			},
+			name:        "empty key",
+			key:         "",
+			data:        "value",
+			expectError: true,
 		},
 	}
 
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			status := FreightStatus{}
-			key := "integration-test-key"
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			status := FreightStatus{
+				Metadata: tc.initialState,
+			}
 
-			// Test Upsert
-			err1 := status.UpsertMetadata(key, tt.data)
-			require.NoError(t, err1)
+			err := status.UpsertMetadata(tc.key, tc.data)
 
-			// Verify metadata was stored
-			assert.NotNil(t, status.Metadata)
-			assert.Contains(t, status.Metadata, key)
+			if tc.expectError {
+				require.Error(t, err)
+				return
+			}
 
-			// Test Get with correct type
-			switch expected := tt.data.(type) {
-			case string:
-				var result string
-				found, err := status.GetMetadata(key, &result)
-				assert.True(t, found)
-				assert.NoError(t, err)
-				assert.Equal(t, expected, result)
+			require.NoError(t, err)
+			require.NotNil(t, status.Metadata)
+			require.Contains(t, status.Metadata, tc.key)
+			if tc.expectedJSON != "" {
+				require.JSONEq(t, tc.expectedJSON, string(status.Metadata[tc.key].Raw))
+			}
+		})
+	}
+}
 
-			case int:
-				var result int
-				found, err := status.GetMetadata(key, &result)
-				assert.True(t, found)
-				assert.NoError(t, err)
-				assert.Equal(t, expected, result)
+func TestFreightStatus_GetMetadata(t *testing.T) {
+	testJSON := map[string]apiextensionsv1.JSON{
+		"string-key": {Raw: []byte(`"test-value"`)},
+		"int-key":    {Raw: []byte(`42`)},
+		"bool-key":   {Raw: []byte(`true`)},
+		"slice-key":  {Raw: []byte(`[1,2,3]`)},
+		"struct-key": {Raw: []byte(`{"name":"test-user","age":30,"active":true}`)},
+	}
 
-			case bool:
-				var result bool
-				found, err := status.GetMetadata(key, &result)
-				assert.True(t, found)
-				assert.NoError(t, err)
-				assert.Equal(t, expected, result)
+	testCases := []struct {
+		name         string
+		metadata     map[string]apiextensionsv1.JSON
```

Comment:
You're using the same metadata in every test case, so there's really no need for `metadata` to be a field of the test case struct.

Reply:
Same applies to the similar tests for Stage metadata retrieval.